File: Arduino/Project/Mobile/Avt.cpp
#include "Avt.h"



void Avt::update(float value){
    if(!this->init){
        this->current_Val = value;
        this->init = true;
        return;
    }
    this->iteration++;
    float error = this->current_Val - value;

    if(error > TOLERENCE){
        this->adjust(UP);
        this->current_Val = max(MIN, this->current_Val - this->delta);
        return;
    }
    if(error < -TOLERENCE){
        this->adjust(DOWN);
        this->current_Val = min(MAX, this->current_Val + this->delta);
        return;
    }

    this->adjust(EQUAL);
    return;
}

void Avt::adjust(int direction){
    if(direction == this->last_direction && direction != EQUAL){
        this->delta *= DELTA_MULTI;
        this->delta = min(this->delta, MAX_DELTA);
    }
    else {
        this->delta /= DELTA_DIV;
        this->delta = max(this->delta, MIN_DELTA);
    }
    this->last_direction = direction;
}

int Avt::get_lastDir(){
    return this->last_direction;
}

Avt::Avt(){
    this->current_Val = MIN;
    this->delta = 1;
    this->iteration = 0;
    this->last_direction = 0;
    // si premiere fois peut-être prendre directement la premiere val
    this->init = false;
}

File: Raspberry_Pi/Server/avt.py
class Avt:

    def __init__(self,mini,maxi,toleranceMin,toleranceMax):
        self.mini = mini
        self.maxi = maxi
        self.currentVal = mini
        self.toleranceMin = toleranceMin
        self.toleranceMax = toleranceMax
        self.delta = toleranceMax
        self.it = 0
        self.sens = 0 # -1 0 ou 1


    def update(self,val):

        error = self.currentVal - val
        self.it+=1

        if error > 0 :
            self.adjust(1)
            self.currentVal = max(self.mini, self.currentVal - self.delta)

        elif error < 0 :
            self.adjust(-1)
            self.currentVal = min(self.maxi, self.currentVal + self.delta)
        else :
            self.adjust(0)

    def adjust(self,sens):

        if sens == self.sens :
            self.delta *= 2
            if self.delta > self.toleranceMax :
                self.delta = self.toleranceMax

        else:
            self.delta *= 1/3.0
            if self.delta < self.toleranceMin :
                self.delta = self.toleranceMin

        self.sens = sens


